### ISSUE TITLE
Packaging similar_posts plugin for PyPi

### ISSUE THREAD

**Issue by kdeldycke** [pelican-plugins/similar-posts#3]
[As of version 4.5](https://blog.getpelican.com/pelican-4.5-released.html), Pelican plugins can now be bundled as independent Python packages and distributed on PyPi. I'd love to have `similar_posts` published on PyPi: it will make maintenance of my site easier!

Are there any plan to package it up?

For the record, Pelican maintainers have created a comprehensive and detailed tutorial at: https://blog.getpelican.com/drafts/namespace-plugin-migration.html

**Comment by davidlesieur** [comment]
Not sure when I'll get to it, but yes, I intend to publish on PyPI.

**Comment by kdeldycke** [comment]
@davidlesieur do you need help? I already migrated the [`neighbors`](https://github.com/pelican-plugins/neighbors), [`thumbnailer`](https://github.com/pelican-plugins/thumbnailer), [`related-posts`](https://github.com/pelican-plugins/related-posts) and [`tipue-search`](https://github.com/pelican-plugins/tipue-search) plugins. I can certainly perform the packaging and publication! 😃 

We can eventually transfer ownership of your plugin to the [`pelican-plugins`](https://github.com/pelican-plugins) umbrella organization, to share the burden of maintenance with the community. What do you think?

**Comment by davidlesieur** [comment]
I don't mind the help! 😃

I have not yet reviewed the migration procedure to the new pelican-plugins organization, but as long as commit history is preserved and I retain write access to the repository and its settings, I'd be good with the move (and wouldn't miss the old pelican-plugins monolith at all!).

**Comment by kdeldycke** [comment]
Hey @justinmayer, do you mind if we let @davidlesieur transfer his repository to us at https://github.com/pelican-plugins ? In the same spirit as you initiated with https://github.com/pelican-plugins/minchin.pelican.plugins.nojekyll ?

That way we can grant @davidlesieur the necessary access right while we work on packaging his plugin?

**Comment by justinmayer** [comment]
Hey David. Glad to hear it. We would of course want you to have full access to the repository post-migration, so no worries there.

One question that came to mind... Do you think it might be sensible to merge the two plugins? (i.e., Related Posts and Similar Posts). I ask because there seems to be a large degree of overlap between them. Plus, there is a collective maintenance burden to keeping plugins up-to-date, so there is incentive to keep the ecosystem from fragmenting unnecessarily.

Even if there were sufficient functional differentiation, it seems to me that a setting could be used to select one algorithm versus the other. That way a single combined plugin could provide the best of the two overlapping plugins' functionality.

What do you think?

**Comment by kdeldycke** [comment]
I'm ready to merge the two if both of you think it is worth the effort.

**Comment by davidlesieur** [comment]
@justinmayer Good thinking here. There is certainly an overlap in the end result provided by the plugins, and a site would probably use either one or the other, but not both plugins. 

However, there is no overlap in code since they work in radically different ways (manual relations in related_posts vs. automatically generated relations in similar_posts). I think a proper merge of the two plugins would allow one not to choose one method vs. the other, but let one enter some relations manually if desired, with the similarity algorithm filling the rest (up to some configured number of relations). That's something I'd be happy to implement.

Another issue is that similar_posts has heavy dependencies (Gensim, NumPy, SciPy). Not a concern to me, but could it be a concern to others, who would have to install those dependencies even when they just want to manually enter some relations?

**Comment by kdeldycke** [comment]
We're onto something here, and I can see great things coming out of our collaboration.

I concur: `gensim` is a fantastic package, but it is huge.  It took me XXX seconds for Poetry to install it on my machine! 😆 But I guess we can add `gensim` and its features optional by the way of an extra `[gensim]` dependency.

Here is the plan I propose:
1. Transfer `similar_posts` to the `pelican-plugins` org. Like, right away.
2. Let me package and publish it on PyPi (should only take a couple of days, I'm used to the process).
3. Continue the discussion about new feature and merging.

That way we have `similar_posts` ready to be used and consumed by the community, while @davidlesieur has the full freedom and liberty to work on the core of the plugin as he see fit.

If we find a good way to merge the two, we can graduate the `similar_posts` as our reference plugin and add a depreciation notice to `related-posts`. Or the other way around depending on the maturity and complexity of each package.

**Comment by davidlesieur** [comment]
@kdeldycke The plan sounds good to me!

**Comment by kdeldycke** [comment]
@davidlesieur Good! So next step is for you to transfer the repository to https://github.com/pelican-plugins . Should be easy: see the red `Transfer` button at the bottom of your repo settings: https://github.com/davidlesieur/similar_posts/settings

**Comment by davidlesieur** [comment]
Based on [GitHub's docs](https://docs.github.com/en/github/administering-a-repository/transferring-a-repository#about-repository-transfers):

> To transfer a repository that you own to an organization, you must have permission to create a repository in the target organization.

It looks like I don't have this permission on pelican-plugins at the moment.

**Comment by davidlesieur** [comment]
By the way, my impression is that following GitHub's process will allow you to simplify the "initial steps" of the current [draft procedure for migrating plugins](https://blog.getpelican.com/drafts/namespace-plugin-migration.html).

**Comment by kdeldycke** [comment]
Oh I see. We can either ask @justinmayer to grant you the necessary permissions, or speed up the process: if you grant me maintainership on your repo I guess I'll be able to transfer it (I have the rights to create new repo on `pelican-plugins`, but can't add new members to the org).

As for the draft procedure, you're right: maybe we should add a note to cover the transfer or an existing repo instead of the creation from scratch.

**Comment by davidlesieur** [comment]
@justinmayer I'd love to be member of the pelican-plugins organization, with permission to create a repo. That would allow me to do the transfer, and help later on other issues. I won't flood the org with new repositories!

**Comment by kdeldycke** [comment]
I can vouch for @davidlesieur 👍 .

And thanks @davidlesieur for the invitation on your repo. Unfortunately the permissions you gave me are not enough to move the repo. 😞 I guess I need a higher role.

**Comment by justinmayer** [comment]
I'm a bit busy this week but will have more time within a week or two. In the mean time, I've invited @davidlesieur to join the Pelican Plugins organization.

@davidlesieur: Looking forward to collaborating with you. If you don't mind, please let me know before creating any other new repositories, as it helps keep things manageable if we can coordinate together regarding when and how that happens. 😸

**Comment by davidlesieur** [comment]
This repository has now been transferred under `pelican-plugins`!

@justinmayer Thanks for your trust. No worries, I won't create new repositories without coordinating with you.

**Comment by kdeldycke** [comment]
Thanks @davidlesieur and @justinmayer for the transfer!

If you're still OK, I'll spend some time this week to prepare a PR with all the packaging work.

**Comment by davidlesieur** [comment]
@kdeldycke Great!

**Comment by davidlesieur** [comment]
@justinmayer Looks like I no longer have access to the repository's settings; can you fix that? Thanks.

**Comment by justinmayer** [comment]
@davidlesieur: Fixed. With organizations + individual repositories, GitHub's permissions are scattered across multiple UIs, which makes for some complicated access management.

On a related note, I changed the underscore in the repo name with a dash in order to maintain consistency with all the other repositories under the new organization. It's nice for users not to have to guess at this sort of thing, which is the benefit of keeping some semblance of consistency. 😊
